src/applications/personalization/view-dependents/manage-dependents/containers/ManageDependentsApp.jsx

```diff
@@ -0,0 +1,116 @@
+import React, { useEffect, useState } from 'react';
+import { connect } from 'react-redux';
+import PropTypes from 'prop-types';
+import SchemaForm from 'platform/forms-system/src/js/components/SchemaForm';
+import LoadingButton from 'platform/site-wide/loading-button/LoadingButton';
+import * as actions from '../redux/actions';
+import { SCHEMAS } from '../schemas';
+
+const ManageDependents = props => {
+  const {
+    relationship,
+    updateFormData,
+    dependentsState,
+    cleanupFormData,
+    closeFormHandler,
+    stateKey,
+  } = props;
+  const [schema, setSchema] = useState(null);
+  const [uiSchema, setUiSchema] = useState(null);
+  const [isFirstLoad, setIsFirstLoad] = useState(true);
+  const onSubmit = () => {};
+
+  const onChange = nextFormData => {
+    updateFormData(
+      dependentsState[stateKey].formSchema,
+      dependentsState[stateKey].uiSchema,
+      nextFormData,
+      stateKey,
+    );
+  };
+
+  const handleFormClose = () => {
+    cleanupFormData(stateKey);
+    closeFormHandler();
+  };
+
+  useEffect(
+    () => {
+      // on first load, setup the things.
+      if (relationship && isFirstLoad) {
+        setIsFirstLoad(false);
+        // grab the schemas needed
+        const initialSchema = SCHEMAS[relationship].schema;
+        const initialUiSchema = SCHEMAS[relationship].uiSchema;
+        // setup initial redux state
+        updateFormData(initialSchema, initialUiSchema, {}, stateKey);
+        // setup local app state
+        setSchema(initialSchema);
+        setUiSchema(initialUiSchema);
+        return;
+      }
+
+      // if dependentsState does exist, we want to use the most recent schemas and uiSchemas from it
+      if (dependentsState[stateKey]) {
+        setSchema(dependentsState[stateKey].formSchema);
+        setUiSchema(dependentsState[stateKey].uiSchema);
+      }
+    },
+    [relationship, dependentsState],
+  );
```

Comment:
Just an idea but I'm looking at that `isFirstLoad` state and thinking there could be a different way to compose this hook. First instinct is a `useEffect` with an empty dependency array so that it only runs on first render. Excuse the whitespace!

```js
const initialize = useCallback(() => {
  if (relationship) {
    // grab the schemas needed
        const initialSchema = SCHEMAS[relationship].schema;
        const initialUiSchema = SCHEMAS[relationship].uiSchema;
        // setup initial redux state
        updateFormData(initialSchema, initialUiSchema, {}, stateKey);
        // setup local app state
        setSchema(initialSchema);
        setUiSchema(initialUiSchema);
}
        
}, [relationship, dependentsState]);

useEffect(() => {
    initialize();
}, []);

useEffect(() => {
     // if dependentsState does exist, we want to use the most recent schemas and uiSchemas from it
      if (dependentsState[stateKey]) {
        setSchema(dependentsState[stateKey].formSchema);
        setUiSchema(dependentsState[stateKey].uiSchema);
      }
}, [dependentsState]);

```

Then I broke out the  logic that uses the `dependentsState` into its own block.

Reply:
Disclaimer - I _think_ `initialize` could just be an ordinary function. I'm not sure if `useCallback` saves anything there.

Reply:
Nice refactor. I knew `isFirstLoad` probably wasn't the best approach, but couldn't see an optimization right away. Thanks for the observation!

Reply:
Of course man! I'm not certain but I think that `onChange` function being wrapped in the `useCallback` will save renders on the `<SchemaForm/>`  because now that prop is less likely to change. Don't worry at all about doing this unless you want to, but you might be able to find out with a little wrapper like -

```jsx
let counter = 0;
function TestingRenders(props) {
    console.log(counter++);
   return <SchemaForm {...props}/>
}
```

Then just swap our `<SchemaForm/>` with that and test first using `useCallback` then again without it to see if there is a difference in render cycles. I did something like that when I was messing with React hooks awhile back. Can actually make a decent performance during a costly operation.